include/deal.II/base/exceptions.h

```diff
@@ -1505,39 +1506,91 @@ namespace deal_II_exceptions
-            __FILE__,                                                    \
-            __LINE__,                                                    \
-            __PRETTY_FUNCTION__,                                         \
-            #cond,                                                       \
-            #exc,                                                        \
-            exc);                                                        \
-      }
-#  else /*ifdef DEAL_II_HAVE_BUILTIN_EXPECT*/
-#    define Assert(cond, exc)                                            \
-      {                                                                  \
-        if (!(cond))                                                     \
-          ::dealii::deal_II_exceptions::internals::issue_error_noreturn( \
-            ::dealii::deal_II_exceptions::internals::ExceptionHandling:: \
-              abort_or_throw_on_exception,                               \
-            __FILE__,                                                    \
-            __LINE__,                                                    \
-            __PRETTY_FUNCTION__,                                         \
-            #cond,                                                       \
-            #exc,                                                        \
-            exc);                                                        \
-      }
-#  endif /*ifdef DEAL_II_HAVE_BUILTIN_EXPECT*/
-#else
+#  if KOKKOS_VERSION >= 30600
+#    ifdef DEAL_II_HAVE_BUILTIN_EXPECT
+#      define Assert(cond, exc)                                              \
+        {                                                                    \
+          KOKKOS_IF_ON_HOST(({                                               \
+            if (__builtin_expect(!(cond), false))                            \
+              ::dealii::deal_II_exceptions::internals::issue_error_noreturn( \
+                ::dealii::deal_II_exceptions::internals::ExceptionHandling:: \
+                  abort_or_throw_on_exception,                               \
+                __FILE__,                                                    \
+                __LINE__,                                                    \
+                __PRETTY_FUNCTION__,                                         \
+                #cond,                                                       \
+                #exc,                                                        \
+                exc);                                                        \
+          }))                                                                \
+          KOKKOS_IF_ON_DEVICE(({                                             \
+            if (!(cond))                                                     \
+              dealii::internal::kokkos_abort(#cond);                         \
+          }))                                                                \
```

Comment:
Another question - can we merge `internal::kokkos_abort()` and `issue_error_noreturn()`? What is preventing us from moving `KOKKOS_IF_ON_HOST()` into that function?

Reply:
The exception variables can't be built on GPUs since they are using `std::string` and virtual functions.